<font color='blue' size=5><b>Комментарий ревьюера</b></font>

<font color='blue'>Привет! Меня зовут Тимофей Вилков, и я буду проверять твой проект. Спасибо за проделанную работу! Предлагаю общаться на «ты».</font>

<font color='blue'>Я буду использовать различные цвета, чтобы было удобнее воспринимать мои комментарии:</font>



<div class="alert alert-block alert-success">
✔️ <b>Зеленый текст</b> - все отлично
</div>

<div class="alert alert-block alert-warning">
⚠️ <b>Желтый текст</b> - сделано все правильно, однако есть рекомендации, на что стоит обратить внимание
</div>

<div class="alert alert-block alert-danger">
❌ <b>Красный текст</b> - красный текст - есть недочеты
</div>

---

<font color='blue'>Пожалуйста, не удаляй мои комментарии в случае возврата работы, так будет проще разобраться, какие были недочеты, а также сразу увидеть исправленное. </font>

Ответы на мои комментарии лучше тоже помечать.
Например:

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Текст комментария, вопроса, объяснения
</div>

<font color='blue'><b>Давай смотреть, что получилось!</b></font>

<div class="alert alert-block alert-success">
✔️ Привет ещё раз! Новые комментарии я отметил так: Ревью №2
</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [37]:
import pandas as pd
from collections import Counter

df = pd.read_csv('/datasets/data.csv')
# смотрю общую информацию по датасету
df.info()
# анализирую таблицу
df.head(20)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Вывод**

Таблица содержит 12 столбцов и 21525 строк. В таблице присутствуют  3 типа данных.
В столбцах days_employed(трудовой стаж) и total_income(ежемесячный доход) по 2174 пропуска. Пропуски находятся в одних строках. Так же тип данных в этих столбцах вещественные числа.
С данными в столбце с трудовым стажем какая то беда произошла. Имеются отрицательные значения и очень большие значения стажа у пенстонеров.
В столбце education записи не в змеином регистре. нужно исправить

<div class="alert alert-block alert-success">
✔️ Отличная работа с первичным выводом! Все главные "недостатки" в данных обнаружены
 </div>

<div class="alert alert-block alert-success">
✔️ Здорово, что обнаружил аномальные значения
</div>

## Шаг 2. Предобработка данных


### Обработка пропусков

In [38]:
#нахожу медиану дохода и стажа соотв.
total_income_m = df['total_income'].median()
days_employed_m = df['days_employed'].median()

# заменяю пропуски медианным значением
df['days_employed'] = df['days_employed'].fillna(days_employed_m)
df['total_income'] = df['total_income'].fillna(total_income_m)

#проверяю что все работает
df.isna().sum()


children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

<div class="alert alert-block alert-success">
✔️ Молоде, что заполняешь пропуски медианой. Но если это делать групппируя по типу занятости, сможем заполнить пропуски точнее. Покажу как можно сделать тоже самое в одну строчку:
<code>df['total_income'].fillna(df.groupby('income_type')['total_income'].transform('median'))</code>
</div>

**Вывод**

Пропуски решил заменить медианным значением т.к. эти столбцы не участвуют в иследовании и по хорошему их можно было удалить.

### Замена типа данных

In [39]:
# замена вещественного типа данных на цельночисленный
df['days_employed'] = df['days_employed'].astype('int')
df['total_income'] = df['total_income'].astype('int')

# привожу столбец education к нижнему регистру
df['education']= df['education'].str.lower()

**Вывод**

Применил метод astype т.к. to_numeric  не подходит, а другого не знаю

<div class="alert alert-block alert-success">
✔️ С типами порядок!
</div>

### Обработка дубликатов

In [40]:


# проблемы со столбцом образование нашел на глаз и перевел в нормальный вид ранее, но на всякий случай посмотрим еще раз
Counter(df['education'])

# проверяю столбец семейное положение
Counter(df['family_status'])
# буду перфикционистом "Не женат" исправлю на "не женат"
df['family_status']= df['family_status'].str.lower()

# проверяю столбец тип занятости
Counter(df['income_type'])

#проверяю столбец цель кредита
Counter(df['purpose'])

#смотрю количество детей
df['children'].value_counts()


# минус исправлю на плюс, вместо 20 подставлю медианное значение 
df['children'] = df['children'].replace(-1, 1)
children_median = df.loc[df.loc[:, 'children'] != 20]['children'].median()
df['children'] = df['children'].replace(20, children_median)

# проверяю что столбец кол-во детей без ошибок
pd.DataFrame({'Кол-во заёмщиков по кол-ву детей':df['children'].value_counts()})

print('дубликатов в таблице', df.duplicated().sum())
print(df['total_income'].value_counts())
df.duplicated(subset = 'total_income').sum()


дубликатов в таблице 71
145017    2175
126262       3
160905       3
150684       3
154199       3
          ... 
109583       1
101387       1
138249       1
280240       1
264193       1
Name: total_income, Length: 18606, dtype: int64


2919

<div class="alert alert-block alert-danger">
❌ Проверь, пожалуйста, наличие полных дубликатов
</div>

Произвел проверку дубликатов. В таблице их много. Например столбец цель кредита полон дубликатов и это логично ведь целей не так много, а желающих взять кредит достаточно. 

<div class="alert alert-block alert-success">
✔️ Ревью №2: Отлично!
</div>

<div class="alert alert-block alert-warning">
⚠️ Импорты стоит собирать в одну ячейку в начале проекта
</div>

Импорты перенес в первую ячейку

<div class="alert alert-block alert-warning">
⚠️ Все выводы не поясняющие работу кода стоит писать markdown
</div>

Выводы убрал из кода

**Вывод**

В дата фрейме были обнаружены ошибки. Ошибки с регистром исправлены. Так же исправлены ошибки с количеством детей. Наверное по хорошему нужно исправить столбец со стажем, но я не понимаю как.#здесь есть ошибки. есть значение -1  детей не может быть меньше 0.  также имеем значение 20 тут либо лишний ноль либо лишняя 2, т.к. наличие у 76 человек по 20 детей маловероятно.

### Лемматизация

In [41]:
'''from pymystem3 import Mystem
m = Mystem()
for i in range(df.shape[0]):
    lemmas = m.lemmatize(df['purpose'][i])
    df.loc[i, 'purpose_lemmas'] = ','.join(lemmas) 
Counter(df['purpose_lemmas'])'''

estate = ['недвижим', 'жиль']
wedding = ['свадьб']
auto = ['автомоб']
education = ['образова']

df['purpose_category'] = 0
def categoryzer(list_of_words, category):
  join = '|'.join(list_of_words)
  index = df[df['purpose'].str.lower().str.contains(join)].index.to_list()
  for i in index:
    df.loc[i, 'purpose_category'] = category
  return df


categoryzer(wedding, 'свадьба')
categoryzer(estate, 'недвижимость')
categoryzer(auto, 'авто')
categoryzer(education, 'образование')

#проверка результата
Counter(df['purpose_category'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
purpose_category    21525 non-null object
dtypes: int64(7), object(6)
memory usage: 2.1+ MB


**Вывод**

Видно что одинаковые цели кредита названы по разному. Выделю 4 категории( образование, недвижимость, свадьба, авто) и добаляю столбец категория цели кредита.

<div class="alert alert-block alert-success">
✔️ Лемматизация проведена успешно. 
</div>

### Категоризация данных

**Вывод**

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [42]:
debt_from_children = pd.DataFrame()
debt_from_children['count_children'] = df.groupby('children')['debt'].count()
debt_from_children['sum_children'] = df.groupby('children')['debt'].sum()
debt_from_children['result_children'] = debt_from_children['sum_children'] / debt_from_children['count_children'] 
debt_from_children.sort_values('result_children', ascending = False)
debt_from_children

count_children  sum_children  result_children
children                                               
0                  14225          1071         0.075290
1                   4865           445         0.091470
2                   2055           194         0.094404
3                    330            27         0.081818
4                     41             4         0.097561
5                      9             0         0.000000

**Вывод**

Поскольку в проекте изучается влияние семейного положения и наличие детей на погашение кредита в срок не буду разбивать возраст на категории, так как они нам не понадобятся.категоризация была применена вместе с лемматизацией выше в одной функции. Категоризовать количество детей нет необходимости, категориями будет их количество, так как всего 6 вариантов.

<div class="alert alert-block alert-danger">
❌ В выводе нет вывода о зависимости между наличием долга и количеством детей) Добавь, пожалуйста
</div>

С увеличением количества детей увеличивается и количество просроченных задолженностей, хотя люди с 3 детьми чаще платят в 
срок чем люди с 1 ребенком. Бездетные реже просрачивают оплату по кредиту.

<div class="alert alert-block alert-success">
✔️ Ревью №2: Огонь!
</div>

In [1]:
- Есть ли зависимость между семейным положением и возвратом кредита в срок?

Object `срок` not found.


In [ ]:
- Есть ли зависимость между семейным положением и возвратом кредита в срок

In [43]:
debt_from_family_status = pd.DataFrame()
debt_from_family_status['sum_family_status'] = df.groupby('family_status')['debt'].sum()
debt_from_family_status['count_family_status'] = df.groupby('family_status')['debt'].count()
debt_from_family_status['result_family_status'] = debt_from_family_status['sum_family_status'] / debt_from_family_status['count_family_status'] 
debt_from_family_status.sort_values('result_family_status', ascending = False)

sum_family_status  count_family_status  \
family_status                                                   
не женат / не замужем                274                 2813   
гражданский брак                     388                 4177   
женат / замужем                      931                12380   
в разводе                             85                 1195   
вдовец / вдова                        63                  960   

                       result_family_status  
family_status                                
не женат / не замужем              0.097405  
гражданский брак                   0.092890  
женат / замужем                    0.075202  
в разводе                          0.071130  
вдовец / вдова                     0.065625

**Вывод**

Зависимость существует. Люди не в браке имеют больший процент невозвратов в срок. Разведенные и овдовевшие чаще платят в срок, чем люди в браке.

<div class="alert alert-block alert-success">
✔️ Есть идеи с чем может быть связано?
</div>

Пологаю что люди которые не в браке менее опытны и способны на спонтанные необдуманные поступки. 
т.к. не несут финансовой ответственности перед семьей(жена, дети) Наверное по большей части это люди молодые. 
Разведенные и овдовевщие скорее всего люди с определенным жизненным багажом и опытом и умеют строить бюджет.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [49]:
#категоризировали уровень дохода создав доп столбец total_income_id. 
def total_income_category(value):
    if value <= 100000:
        return 'от 0 до 100'
    elif value > 100000 and value <= 200000:
        return 'от 100 до 200'
    elif value > 200000 and value <= 300000:
        return 'от 200 до 300'
    elif value > 300000 and value <= 400000:
        return 'от 300 до 400'
    else:
        return 'больше 400'
df['total_income_id'] = df['total_income'].apply(total_income_category)
df['total_income_id'].value_counts()

debt_from_total_income = pd.DataFrame()
debt_from_total_income['sum'] = df.groupby('total_income_id')['debt'].sum()
debt_from_total_income['count'] = df.groupby('total_income_id')['debt'].count()
debt_from_total_income['conversion'] = debt_from_total_income['sum'] / debt_from_total_income['count']
print(debt_from_total_income.sort_values('conversion', ascending = False))




#видим что зависимости долга от доходов кажется нет, на всякий случай посмотрим корреляцию между столбцами дохода и долгом
verification_through_correlation = pd.DataFrame()
verification_through_correlation['total_income'] = df['total_income']
verification_through_correlation['debt'] = df['debt']
verification_through_correlation.corr()
#видим, что корреляции нет, теперь вывод очевиден

                  sum  count  conversion
total_income_id                         
от 100 до 200    1029  11996    0.085779
от 0 до 100       354   4463    0.079319
от 300 до 400      75    954    0.078616
от 200 до 300     252   3584    0.070312
больше 400         31    528    0.058712


total_income      debt
total_income      1.000000 -0.011592
debt             -0.011592  1.000000

**Вывод**

<div class="alert alert-block alert-danger">
❌ Категоризируй доход на несколько категорий и проверь наличие зависимости ещё раз 
</div>

<div class="alert alert-block alert-success">
✔️ Ревью №2: Теперь порядок!
</div>

Разделил доход на 5 категорий, похоже что зависимость уровня дохода и возврата кредита в срок не очевидна.

- Как разные цели кредита влияют на его возврат в срок?

In [46]:
debt_from_purpose_category = pd.DataFrame()
debt_from_purpose_category['sum_purpose_category'] = df.groupby('purpose_category')['debt'].sum()
debt_from_purpose_category['count_purpose_category'] = df.groupby('purpose_category')['debt'].count()
debt_from_purpose_category['result_purpose_category'] = debt_from_purpose_category['sum_purpose_category'] / debt_from_purpose_category['count_purpose_category'] 
debt_from_purpose_category.sort_values('result_purpose_category', ascending = False)

sum_purpose_category  count_purpose_category  \
purpose_category                                                 
авто                               403                    4315   
образование                        370                    4022   
свадьба                            186                    2348   
недвижимость                       782                   10840   

                  result_purpose_category  
purpose_category                           
авто                             0.093395  
образование                      0.091994  
свадьба                          0.079216  
недвижимость                     0.072140

**Вывод**

среди 4  категорий, чаще всего платят в срок с целью кредита недвижимость, чаще просрочка у тех кому нужен кредит на авто и образование. Кредиты на свадьбу просрачивают реже - примерно как и в случае недвижимости.

## Шаг 4. Общий вывод

Cемейный статус влияет на вероятность платежей по кредиту в срок - люди в браке чаще плятят в срок чем те, кто не браке. Разведенные и овдовевшие платят в срок чаще, чем люди в браке. Чем больше детей, тем чаще платят не в срок, бездетные реже просрачивают оплату.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

<font color='blue'><b>Итоговый комментарий ревьюера</b></font>\
<font color='green'>Андрей, хороший проект, видно, что ты можешь много показать, нужно исправить лишь пару нюансов:)
   

<font color='blue'>Что можно сделать лучше:</font>
<ul><font color='red'>1) Проверить полные дубликаты </font></ul>
<ul><font color='red'>2) Добавить гипотезы в ответы на вопросы там, где их нет</font></ul>
<ul><font color='red'>3) Категоризировать доход</font></ul>
    
<font color='blue'><b>Спасибо, жду твоих исправлений и комментариев!</b></font>

<div class="alert alert-block alert-success">
✔️ Ревью №2: Андрей, спасибо за внесённые правки, работа стала ещё лучше! Перевожу твой проект в статус "зачтено" и желаю дальнейших успехов в освоении курса
</div>